### DISCUSSION TITLE
Passing Data Ids to frontend is Good Practice Or Not ?

### DISCUSSION THREAD

**Discussion by EhanDuan** [vercel/next.js#37898] [Help] [answered]
Hi, I have a question about passing data b/w **frontend** and **backend**.

---
### Scenario
My question is from the following scenario: 
1. fetching data from database, displaying to `users` on webpage
2. users interact (CRUD)
3. sending interaction data back to database

### My thoughts
For **users**, actual data fields like `name` is meaningful to them. So I need to display `name`  and other fields.

For **server or me developer**, the `id` of that piece of data is important since I can directly access to that doc and then update. Like firestore, we use `id` to access. Even using query where name == [name], the purpose is still to get `id`.

So in order to meet the above 2 needs, I decided to get both `ids` and `datas` together, like below where I fetched data at request time
```tsx
export async function getServerSideProps() {
  // Fetch data
  const querySnapshot = await getDocs(collection(db, "skills"));
  const skills: any[] = [];
  const sIds: any[] = [];

  querySnapshot.forEach((doc) => {
    skills.push(doc.data());
    sIds.push(doc.id);
  });

  // Pass data to the page via props
  return { props: { skills, sIds } };
}
```

#### Question 1
**Is it a good practice to pass doc ids to the frontend?** Will it cause `ids` exposed to users that might cause safety concern? **If it is not a good one, how do I handle this scenario?**

#### Question 2
`getServerSideProps` only allows serializable(json) object. So I cannot directly generate a hashmap (id to name) at request time. 

Since I need the hashmap, I will create one in the frontend ( Like below code. If it is not frontend, please point it out. I am not so familar with **code elimination** rule).

For each page or component like this, there would be hashmap for each. **Will this cause a burden at frontend side?**

In case you need, here is the component code.
```tsx
const FocusList = ({ skills, sIds }) => {
 // some process....

  // create a roaster (mapping relationship b/w sId and skill name)
  // Nextjs props need to be serialized like json. So I cannot generate this hashmap directly in getServerSideProps and pass.
  const roaster = new Map<string, string>();
  for (let i = 0; i < sIds.length; i++) {
    roaster.set(sIds[i], skills[i].name);
  }

  return (
        <CheckboxGroup >
          <Stack spacing={[1, 5]} direction={["column", "row"]}>
            {/* I want to display skills' names and hidden skills' ids which are used for db */}
            {sIds.length > 0 &&
              sIds.map((sId, index) => (
                <Checkbox
                  key={index}
                  value={sId}
                  onChange={handleCheck}
                  isChecked={selectedMap.get(sId)}
                >
                  {roaster.get(sId)}
                </Checkbox>
              ))}
          </Stack>
        </CheckboxGroup>
  );
```

**Comment by icyJoseph** [comment, marked as answer]
> Is it a good practice to pass doc ids to the frontend? Will it cause ids exposed to users that might cause safety concern? If it is not a good one, how do I handle this scenario?

Well, id's should be opaque data, meaning that, you gain nothing out of having it. 

Sometimes, we design a system that relies on this id's for actionable purposes, and in such cases, we might not want to reveal the real id. 

So it depends on the application. Asking for best practices is both positive and negative. Of course, we don't want to do things which are blatantly wrong, but at the end of the day, these subjects ought to be analysed on a "per-case" base.

I once build a Battlesnake app, and stored the game history on Supabase. An to be honest, for whatever reason, I didn't want to expose the game UUID, to the client, so I encrypted it.

```js
 const encrypted = data
    .map(({ uuid, moves, start_game, ...rest }) => {
      const { iv, content } = encrypt(uuid);

      return {
        ...rest,
        snake_name: start_game.you.name,
        pk: iv,
        uuid: content,
        total_moves: moves.length
      };
    })
```

Where the `encrypt/decrypt` pair  looks like:

```ts
import crypto from "crypto";

const algorithm = "aes-256-ctr";
const secretKey = process.env.STATS_SECRET_KEY;
const iv = crypto.randomBytes(16);

const encrypt = (text: string) => {
  const cipher = crypto.createCipheriv(algorithm, secretKey, iv);

  const encrypted = Buffer.concat([cipher.update(text), cipher.final()]);

  return {
    iv: iv.toString("hex"),
    content: encrypted.toString("hex")
  };
};

const decrypt = (hash: { iv: string; content: string }) => {
  const decipher = crypto.createDecipheriv(
    algorithm,
    secretKey,
    Buffer.from(hash.iv, "hex")
  );

  const decrpyted = Buffer.concat([
    decipher.update(Buffer.from(hash.content, "hex")),
    decipher.final()
  ]);

  return decrpyted.toString();
};

export { encrypt, decrypt };
```

This of course becomes one extra pass over the list, encrypting every game unique id, and also adds the need to decrypt when trying to query a game, but end users could not ever get to know the game id from my client.

```ts
 const { data, error } = await supabase
    .from<{ uuid: string }>("battlesnake_history")
    .select(
      "uuid,has_ended,created_at,ended_at,start_game,end_game,moves,winner"
    )
    .eq("uuid", decrypt({ iv: pk, content: uuid }))
    .single();
```

So, it depends on your use case, but itself is not a security risk, it'd need some kind of architecture design that turns it into a risk.

> getServerSideProps only allows serializable(json) object. So I cannot directly generate a hashmap (id to name) at request time.
>
> Since I need the hashmap, I will create one in the frontend ( Like below code. If it is not frontend, please point it out. I am not so familar with code elimination rule).
>
> For each page or component like this, there would be hashmap for each. Will this cause a burden at frontend side?

 I think it is fine to reconstruct the Map client side. I prefer maps over objects, but it feels like in this case, perhaps having the same data structure in both ends would be good?

This is an example of hydrating a data structure, given a dehydrated input.

One thing I'd do though:

```ts
  const roaster = useMemo(
        () =>
            sIds.reduce<Map<string, string>>((prev, current, index) => {
                prev.set(current, skills[index].name);

                return prev;
            }, new Map()),
        [sIds, skills]
    );
```

or

```ts
   const roaster = useMemo(() => {
        const map = new Map<string, string>();

        sIds.forEach((current, index) => {
            map.set(current, skills[index].name);
        });

        return map;
    }, [sIds, skills]);
```

Which ever flavor you prefer, even with the good ol' for loop, but the idea is to use the `useMemo` hook to not run the map creation every time the component does a render pass. 

I could throw a comment on the need of the Map though. And also the use of `key={index}`, [which has its own problems](https://react-fundamentals.netlify.app/isolated/final/07.extra-1.js), click on a field on each case, and see how it behaves.

```jsx

 {sIds.length > 0 &&
              sIds.map((sId, index) => (
                <Checkbox
                  key={sid}
                  value={sId}
                  onChange={handleCheck}
                  isChecked={selectedMap.get(sId)}
                >
                  {skills[index].name}
                </Checkbox>
              ))}
```

Wouldn't just doing, `{skills[index].name}` be the same as looking through the Map? It uses less memory, and less computing time, because you just read from the `skills` and `sIds`. I am not sure that you are getting any benefits out of the Map, the way you have it anyway, because you still map over `sIds`, which might have duplicates, perhaps that's the reason for the index usage?

With the Map, you could do, `Array.from(roaster.entries()).map(([sId, skillName]) => {})`, optionally, you could memoize the result of `Array.from(roaster.entries())`

**Reply by EhanDuan** [reply]
@icyJoseph Hi icy, thanks for your super detailed explanation. It completely solved my questions. I feel very warm in this community. Especially you also pointed out my other errors. I learned a lot here!
